syntax = "proto3";

package vtserver_interface;

// common structure for all response messages
message requestResult {
  bool success = 1;
  string msg = 2; // error or info message
}

message Timestamp {
  // https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto
  int64 seconds = 1;
  int32 nanos = 2;
}

// ---------------------------------
// ---------- Dataset API ----------
// ---------------------------------

message datasetInfo {
  string dataset_id = 1;
  string name = 2;
  string friendly_name = 3;
  string description = 4;
}

message datasetMetrics {
  string dataset_id = 1;
  int64 video_count = 2;
  int64 process_count = 3;
  int64 task_count = 4;
}

// addDataset
message addDatasetRequest {
  string name = 1;
  string friendly_name = 2;
  string description = 3;
}

message addDatasetResponse {
  requestResult res = 1;
  string dataset_id = 2;
}

// getDatasetList
message getDatasetListRequest {
}

message getDatasetListResponse {
  requestResult res = 1;
  repeated datasetInfo datasets = 2;
}

// getDatasetMetrics(string #datasetID) → dataset_metrics metrics
message getDatasetMetricsRequest {
  string dataset_id = 1;
}

message getDatasetMetricsResponse {
  requestResult res = 1;
  datasetMetrics metrics = 2;
}

// deleteDataset (string #datasetID) → bool success
message deleteDatasetRequest {
  string dataset_id = 1;
}

message deleteDatasetResponse {
  requestResult res = 1;
}

// ---------------------------------
// ---------- Videos API -----------
// ---------------------------------

// TODO: oneof?
message videoInfo {
  string video_id = 1;
  string filepath = 2; // complete path to the file
  string location = 3; // physical location
  Timestamp start_time = 4; // real-world video start time
  string comment = 5;
  int64 length_frames = 6; // # of frames in video
  double length_ms = 7; // video length in ms
  double fps = 8; // frames per second
  double speed = 9; // 1.0 = normal video speed
  Timestamp added_time = 10; // time the video was added to dataset
}

// addVideo(string  #datasetID, string filepath, float speed, timestamp start_time, string comment, string location) → string #videoID
message addVideoRequest {
  string dataset_id = 1;
  string filepath = 2; 
  string name = 3;
  string location = 4;
  Timestamp start_time = 5;
  double speed = 6;
  string comment = 7;
}

message addVideoResponse {
  requestResult res = 1;
  string video_id = 2;
}

// postVideo (string #datasetID,blob video, float speed, timestamp start_time, string comment) → string #videoID
// TBD

// getVideoIDList (string #datasetID) → string #videoIDs[]
message getVideoIDListRequest {
  string dataset_id = 1;
}

message getVideoIDListResponse {
  requestResult res = 1;
  repeated string video_ids = 2;
}

// getVideoInfo (string #datasetID, string #videoIDs[]) → video_info videos[]
message getVideoInfoRequest {
  string dataset_id = 1;
  repeated string video_ids = 2;
}

message getVideoInfoResponse {
  requestResult res = 1;
  repeated videoInfo videos = 2;
}

//setVideoInfo (string #datasetID, string #videoID, timestamp start_time) → bool success
message setVideoInfoRequest {
  string dataset_id = 1;
  string video_id = 2;
  Timestamp start_time = 3;
}

message setVideoInfoResponse {
  requestResult res = 1;
}

// deleteVieo (string #datasetID, string #videoID)
message deleteVideoRequest {
  string dataset_id = 1;
  string video_id = 2;
}

message deleteVideoResponse {
  requestResult res = 1;
}

// ---------------------------------
// ----- Processing tasks API ------
// ---------------------------------

message taskParam {
  enum taskParamType {
  TP_STRING = 0;
  TP_INT = 1;
  TP_INTARRAY = 2;
  TP_FLOAT = 3;
  TP_FLOATARRAY = 4;
  }
  taskParamType type = 1;
  string name = 2;
  string value_string = 3;
  int64 value_int = 4;
  repeated int64 value_int_array = 5;
  double value_float = 6;
  repeated double value_float_array = 7;
}

message taskInfo {
  string task_id = 1;
  string module = 2;
  repeated taskParam params = 3;
  string prereq_task_id = 4;
  repeated string process_ids = 5;
  Timestamp added_time = 6;
}

message taskProgress {
  double progress = 1; // 0-1
  Timestamp time_to_finish = 2;
  repeated string inprogress_video_ids = 3;
  repeated string done_video_ids = 4;
}

// addTask (string #datasetID, string module, string prereq_task_id, task_param params[]) → string #taskID
message addTaskRequest {
  string dataset_id = 1;
  string module = 2;
  string prereq_task_id = 3;
  repeated taskParam params = 4;
}

message addTaskResponse {
  requestResult res = 1;
  string task_id = 2;
}

// getTaskIDList (string #dataset) → string #taskIDs[]
message getTaskIDListRequest {
  string dataset_id = 1;
}

message getTaskIDListResponse {
  requestResult res = 1;
  repeated string task_ids = 2;
}

// getTaskInfo (string #datasetID, string #taskID[]) → task_info tasks[]
message getTaskInfoRequest {
  string dataset_id = 1;
  repeated string task_ids = 2;
}

message getTaskInfoResponse {
  requestResult res = 1;
  repeated taskInfo tasks = 2;
}

// getTaskProgress(string #datasetID, string #taskID, string #videoID[]) → task_progress
message getTaskProgressRequest {
  string dataset_id = 1;
  string task_id = 2;
  repeated string video_ids = 3;
}

message getTaskProgressResponse {
  requestResult res = 1;
  taskProgress task_progress = 2;
}

// deleteTask (string #datasetID, string #taskID, bool force_data, bool force_dependencies) → bool success
message deleteTaskRequest {
  string dataset_id = 1;
  string task_id = 2;
  bool force_data = 3; // removes all computed data associated with the task
  bool force_dependencies = 4; // also removes all tasks dependent on this one
}

message deleteTaskResponse {
  requestResult res = 1;
}

// ---------------------------------
// -------- Processes API ----------
// ---------------------------------

message processInfo {
  string process_id = 1;
  string assigned_task_id = 2; // which task is being computed
  repeated string assigned_video_ids = 3; // which videos are being processed
  enum processState {
  STATE_CREATED = 0;
  STATE_RUNNING = 1;
  STATE_FINISHED = 2;
  STATE_ERROR = 3;
  }
  processState state = 4; // current state
  double progress = 5; // 0-100
  string current_item = 6; // currently processed video
  string error_message = 7; // error message on STATE_ERROR state
  Timestamp added_time = 8; // time the process was added to dataset
}

// getProcessIDList(string #datasetID, string #module) → string #processIDs[]
message getProcessIDListRequest {
  string dataset_id = 1;
  string module = 2;
}

message getProcessIDListResponse {
  requestResult res = 1;
  repeated string process_ids = 2;
}

// getProcessInfo (string #datasetID, string #processID[]) → process_info processes[]
message getProcessInfoRequest {
  string dataset_id = 1;
  repeated string process_ids = 2;
}

message getProcessInfoResponse {
  requestResult res = 1;
  repeated processInfo processes = 2;
}

// runProcess (string #datasetID, string #videoIDs[], string #taskID) → string #processID
message runProcessRequest {
  string dataset_id = 1;
  repeated string video_ids = 2;
  string task_id = 3;
}

message runProcessResponse {
  requestResult res = 1;
  string process_id = 2;
}

// stopProcess (string #datasetID, string #processID) → bool success
message stopProcessRequest {
  string dataset_id = 1;
  string process_id = 2;
}

message stopProcessResponse {
  requestResult res = 1;
}

// ---------------------------------
// -------- Events API -------------
// ---------------------------------

message Region {
  int64 t = 1;
  double t_sec = 2;
  double x1 = 3;
  double x2 = 4;
  double y1 = 5;
  double y2 = 6;
}

message eventInfo {
  int64 t1 = 1; // (frames)
  int64 t2 = 2; // (frames)
  double t1_sec = 3; // (seconds)
  double t2_sec = 4; // (seconds)
  double length = 5; // (seconds)
  int64 group_id = 6; 
  int64 class_id = 7;
  double score = 8;
  repeated Region regions = 9; // list of bounding boxes - trajectory
}

message eventInfoList {
  string video_id = 1;
  repeated eventInfo events = 2;
}

message eventStats {
  string video_id = 1;
  int64 count = 2;
  double coverage = 3;
  bytes coverage_bitmap = 4;
}

// TODO: oneof?
message eventFilter {
  double min_duration = 1;
  double max_duration = 2;
  Timestamp begin_timewindow = 3;
  Timestamp end_timewindow = 4;
  Timestamp begin_daywindow = 5;
  Timestamp end_daywindow = 6;
  Region region = 7;
}

// getEventList (string #datasetID, string #videoIDs[], string #taskID, event_filter filter) → event_info_list events_list[]
message getEventListRequest {
  string dataset_id = 1;
  repeated string video_ids = 2;
  string task_id = 3;
  eventFilter filter = 4;
}

message getEventListResponse {
  requestResult res = 1;
  repeated eventInfoList events_list = 2;
}

// getEventsStats string #datasetID, string #videoIDs[], string #taskID, event_filter filter, bool get_bitmap) → event_stats stats[]
message getEventsStatsRequest {
  string dataset_id = 1;
  repeated string video_ids = 2;
  string task_id = 3;
  eventFilter filter = 4;
  bool get_bitmap = 5;
}

message getEventsStatsResponse {
  requestResult res = 1;
  repeated eventStats stats = 2;
}


// ---------------------------------
// -- VideoProcessing metadata API -
// ---------------------------------

message classIdOccurence {
  int64 class_id = 1;
  double occurrence = 2;
}

message processingMetadataVideoType {
  repeated classIdOccurence class_id_occurence= 1;
}

message getProcessingMetadataRequest {
  string dataset_id = 1;
  repeated string video_ids = 2;
  string task_id = 3;
}

message getProcessingMetadataResponse {
  requestResult res = 1;
  processingMetadataVideoType metadata_videotype = 2;
}


// ---------------------------------
// ----- RPC service definition ----
// ---------------------------------

service VTServerInterface {
  rpc addDataset(addDatasetRequest) returns(addDatasetResponse);
  rpc getDatasetList(getDatasetListRequest) returns(getDatasetListResponse);
  rpc getDatasetMetrics(getDatasetMetricsRequest) returns(getDatasetMetricsResponse);
  rpc deleteDataset(deleteDatasetRequest) returns(deleteDatasetResponse);
  rpc addVideo(addVideoRequest) returns(addVideoResponse);
  rpc getVideoIDList(getVideoIDListRequest) returns(getVideoIDListResponse);
  rpc getVideoInfo(getVideoInfoRequest) returns(getVideoInfoResponse);
  rpc setVideoInfo(setVideoInfoRequest) returns(setVideoInfoResponse);
  rpc deleteVideo(deleteVideoRequest) returns(deleteVideoResponse);
  rpc addTask(addTaskRequest) returns(addTaskResponse);
  rpc getTaskIDList(getTaskIDListRequest) returns(getTaskIDListResponse);
  rpc getTaskInfo(getTaskInfoRequest) returns(getTaskInfoResponse);
  rpc getTaskProgress(getTaskProgressRequest) returns(getTaskProgressResponse);
  rpc deleteTask(deleteTaskRequest) returns(deleteTaskResponse);
  rpc getProcessIDList(getProcessIDListRequest) returns(getProcessIDListResponse);
  rpc getProcessInfo(getProcessInfoRequest) returns(getProcessInfoResponse);
  rpc runProcess(runProcessRequest) returns(runProcessResponse);
  rpc stopProcess(stopProcessRequest) returns(stopProcessResponse);
  rpc getEventList(getEventListRequest) returns(getEventListResponse);
  rpc getEventsStats(getEventsStatsRequest) returns(getEventsStatsResponse);
  rpc getProcessingMetadata(getProcessingMetadataRequest) returns(getProcessingMetadataResponse);
}

